File: apps/mobile/src/lib/runtimeModeMenu.ts
import type { RuntimeMode } from "@t3tools/contracts";
import { HERMES_RUNTIME_MODE_CHOICES, hermesRuntimeModeChoice } from "@t3tools/shared/runtimeModes";

export interface RuntimeModeMenuOption {
  readonly mode: RuntimeMode;
  readonly title: string;
  /** One word for the composer's compact settings summary pill. */
  readonly shortLabel: string;
}

/** Mobile labels the first mode "Approve actions" rather than web's "Supervised". */
const DEFAULT_RUNTIME_MODE_MENU_OPTIONS: ReadonlyArray<RuntimeModeMenuOption> = [
  { mode: "approval-required", title: "Approve actions", shortLabel: "Approve" },
  { mode: "auto-accept-edits", title: "Auto-accept edits", shortLabel: "Edits" },
  { mode: "auto", title: "Auto", shortLabel: "Auto" },
  { mode: "full-access", title: "Full access", shortLabel: "Full" },
];

const HERMES_RUNTIME_MODE_MENU_OPTIONS: ReadonlyArray<RuntimeModeMenuOption> =
  HERMES_RUNTIME_MODE_CHOICES.map((choice) => ({
    mode: choice.mode,
    title: choice.label,
    // Hermes' two choices are the approve/full pair, so they summarize the
    // same way the generic modes of those names do.
    shortLabel: choice.mode === "full-access" ? "Full" : "Approve",
  }));

/**
 * The Runtime menu entry for a thread. T3 Work (Hermes) threads get the two
 * options Hermes actually distinguishes; everything else gets the four generic
 * modes. A mode outside the offered set still has to read as something, since a
 * thread can carry one in from wherever it was created.
 */
export function runtimeModeMenu(input: {
  readonly isHermes: boolean;
  readonly runtimeMode: RuntimeMode;
}): {
  readonly options: ReadonlyArray<RuntimeModeMenuOption>;
  readonly selected: RuntimeModeMenuOption;
} {
  if (input.isHermes) {
    const mode = hermesRuntimeModeChoice(input.runtimeMode).mode;
    return {
      options: HERMES_RUNTIME_MODE_MENU_OPTIONS,
      selected: HERMES_RUNTIME_MODE_MENU_OPTIONS.find((option) => option.mode === mode)!,
    };
  }
  return {
    options: DEFAULT_RUNTIME_MODE_MENU_OPTIONS,
    selected:
      DEFAULT_RUNTIME_MODE_MENU_OPTIONS.find((option) => option.mode === input.runtimeMode) ??
      DEFAULT_RUNTIME_MODE_MENU_OPTIONS[0]!,
  };
}

File: apps/swift-ios/Features/Workspace/RuntimeModeMenu.swift
import Foundation

// Ported from apps/mobile/src/lib/runtimeModeMenu.ts, together with the
// HERMES_RUNTIME_MODE_CHOICES it reads out of packages/shared/src/runtimeModes.ts.

/// One of the access options a T3 Work (Hermes) thread offers.
///
/// Hermes runs its own gate on the commands it considers dangerous, and its
/// session protocol carries no approval or sandbox setting to relax it, so the
/// four generic modes collapse to the two things T3 can actually do with an
/// `approval.request`: answer it, or show it. Supervised, auto-accept edits and
/// auto would all behave identically here, so the picker offers this pair rather
/// than three labels promising distinctions Hermes does not make.
public struct HermesRuntimeModeChoice: Equatable, Sendable {
    public let mode: RuntimeMode
    public let label: String
    public let description: String
}

public enum HermesRuntimeModes {
    public static let choices: [HermesRuntimeModeChoice] = [
        HermesRuntimeModeChoice(
            mode: .approvalRequired,
            label: "Approve risky commands",
            description: "Hermes asks before running anything it flags as dangerous."
        ),
        HermesRuntimeModeChoice(
            mode: .fullAccess,
            label: "Full access",
            description: "Let Hermes run commands without asking."
        ),
    ]

    /// The choice a Hermes thread's stored mode displays as. A thread can carry
    /// `auto` or `autoAcceptEdits` in from wherever it was created; both ask on
    /// Hermes, so both read as the approval choice rather than as a label the
    /// picker does not offer.
    public static func choice(for mode: RuntimeMode) -> HermesRuntimeModeChoice {
        choices.first { $0.mode == mode } ?? choices[0]
    }
}

public struct RuntimeModeMenuOption: Equatable, Sendable {
    public let mode: RuntimeMode
    public let title: String
}

/// The Runtime menu entry for a thread.
public struct RuntimeModeMenu: Equatable, Sendable {
    public let options: [RuntimeModeMenuOption]
    public let selected: RuntimeModeMenuOption

    /// Mobile labels the first mode "Approve actions" rather than web's
    /// "Supervised".
    private static let defaultOptions: [RuntimeModeMenuOption] = [
        RuntimeModeMenuOption(mode: .approvalRequired, title: "Approve actions"),
        RuntimeModeMenuOption(mode: .autoAcceptEdits, title: "Auto-accept edits"),
        RuntimeModeMenuOption(mode: .auto, title: "Auto"),
        RuntimeModeMenuOption(mode: .fullAccess, title: "Full access"),
    ]

    private static let hermesOptions: [RuntimeModeMenuOption] = HermesRuntimeModes.choices.map {
        RuntimeModeMenuOption(mode: $0.mode, title: $0.label)
    }

    /// T3 Work (Hermes) threads get the two options Hermes actually
    /// distinguishes; everything else gets the four generic modes. A mode
    /// outside the offered set still has to read as something, since a thread
    /// can carry one in from wherever it was created.
    public static func resolve(isHermes: Bool, runtimeMode: RuntimeMode) -> RuntimeModeMenu {
        if isHermes {
            let mode = HermesRuntimeModes.choice(for: runtimeMode).mode
            return RuntimeModeMenu(
                options: hermesOptions,
                selected: hermesOptions.first { $0.mode == mode } ?? hermesOptions[0]
            )
        }
        return RuntimeModeMenu(
            options: defaultOptions,
            selected: defaultOptions.first { $0.mode == runtimeMode } ?? defaultOptions[0]
        )
    }
}
